### PR TITLE
Get docker-compose from developer-script repo

### PR DESCRIPTION
**Pull request by jinlinGuan**
issue #295 

variable RELEASE to select specific release, the default is nightly-build
variable FOR_REDIS=ture to select redis-no-secty.yml
variable SECURITY_SERVICE_NEEDED=true to select security compose file

It successfully run on my local(macOS Catalina), but gets errors on jenkins sandbox.
Running core data test:
image

Signed-off-by: Ginny Guan <ginny@iotechsys.com>

### CODE REVIEW

**Comment by cloudxxx8** [comment]
This PR is working in process.  We can run on local machine, but encounter issue on Sandbox.
We need more time to resolve it.

**Comment by cloudxxx8** [comment]
We have resolved the permission issue, and it could pass on sandbox now

**Comment by cloudxxx8** [comment]
@jinlinGuan please rebase and resolve the conflict

File: bin/arm64_env.sh [hunk 1]
```diff
@@ -1,27 +1,4 @@
...
-export coreMetadata=nexus3.edgexfoundry.org:10004/docker-core-metadata-go-arm64:master
-export coreCommand=nexus3.edgexfoundry.org:10004/docker-core-command-go-arm64:master
-
-export supportLogging=nexus3.edgexfoundry.org:10004/docker-support-logging-go-arm64:master
-export supportNotifications=nexus3.edgexfoundry.org:10004/docker-support-notifications-go-arm64:master
-export supportScheduler=nexus3.edgexfoundry.org:10004/docker-support-scheduler-go-arm64:master
-export supportRulesengine=nexus3.edgexfoundry.org:10004/docker-support-rulesengine-arm64:master
-
-export deviceVirtual=nexus3.edgexfoundry.org:10004/docker-device-virtual-go-arm64:master
 export appService=nexus3.edgexfoundry.org:10004/docker-app-service-configurable-arm64:latest
```
Comment (lranjbar) [inline bin/arm64_env.sh]: Why isn't this consistent with the rest of the services? This should be pulled from the compose file as well, correct?

File: bin/env.sh [hunk 2]
```diff
@@ -1,28 +1,5 @@
...
-export systemManagement=nexus3.edgexfoundry.org:10004/docker-sys-mgmt-agent-go:master
-
-export vault=vault:1.0.3
-export vaultWorker=nexus3.edgexfoundry.org:10004/docker-edgex-security-secretstore-setup-go:master
-export kongdb=postgres:9.6
-export kong=kong:1.3.0-centos
-export edgexProxy=nexus3.edgexfoundry.org:10004/docker-edgex-security-proxy-setup-go:master
-export securitySecretsSetup=nexus3.edgexfoundry.org:10004/docker-edgex-secrets-setup-go:master
-
 export postman=postman/newman
```
Comment (lranjbar) [inline bin/env.sh]: This version of newman should be pinned to same specific version identifed in the ARM build.

File: bin/env.sh [hunk 3]
```diff
@@ -1,28 +1,5 @@
...
-export coreMetadata=nexus3.edgexfoundry.org:10004/docker-core-metadata-go:master
-export coreCommand=nexus3.edgexfoundry.org:10004/docker-core-command-go:master
-
-export supportLogging=nexus3.edgexfoundry.org:10004/docker-support-logging-go:master
-export supportNotifications=nexus3.edgexfoundry.org:10004/docker-support-notifications-go:master
-export supportScheduler=nexus3.edgexfoundry.org:10004/docker-support-scheduler-go:master
-export supportRulesengine=nexus3.edgexfoundry.org:10004/docker-support-rulesengine:master
-
-export deviceVirtual=nexus3.edgexfoundry.org:10004/docker-device-virtual-go:master
 export appService=nexus3.edgexfoundry.org:10004/docker-app-service-configurable:latest
```
Comment (lranjbar) [inline bin/env.sh]: Also should be pulled from the compose file.

**Review by lranjbar — CHANGES_REQUESTED** [review]

**Comment by bnevis-i** [comment]
@jinlinGuan I am diffing developer-scripts and blackbox-test and noticed that blackbox has /etc/newman mounted to several places.  Can you plain this more? What is this volume mount used for?

**Comment by cloudxxx8** [comment]
@bnevis-i It only exists in volume and mongo as I see.  There are some historical reasons, but they are useless right now.  This PR will eliminate those /etc/newman folder from other containers.

File: bin/arm64_env.sh (hunk 1 above)
Comment (cloudxxx8) [reply bin/arm64_env.sh]: @lranjbar this is a special case.  There are two application services running for blackbox test.  One is **app-service-rules** and another is **app-service-configurable**.  They are using the same docker image with different configuration because it is **configurable** application service.  The app-service-configurable here is provided for test only, so it doesn't exist in developer-script.
https://github.com/edgexfoundry/developer-scripts/blob/master/releases/nightly-build/compose-files/docker-compose-nexus.yml

You can see the configuration profile here is `blackbox-tests`

```
  app-service-configurable:
    image: ${appService}
    ports:
      - "48095:48095"
    entrypoint: ["/app-service-configurable"]
    command: ["--registry","--confdir=/res", "--profile=blackbox-tests", "--skipVersionCheck=true"]
    container_name: edgex-app-service-configurable
    hostname: app-service-configurable
    environment:
      edgex_registry: consul://edgex-core-consul:8500
      edgex_service: http://app-service-configurable:48095
      edgex_profile: blackbox-tests
      Service_Host: app-service-configurable
      Clients_CoreData_Host: edgex-core-data
      Clients_Logging_Host: edgex-support-logging
      Logging_EnableRemote: "true"
      MessageBus_SubscribeHost_Host: edgex-core-data
      Database_Host: edgex-mongo
      Database_Username: appservice
      Database_Password: password
    networks:
      edgex-network:
        aliases:
          - edgex-app-service-configurable
```

Thus, we need to keep this part in this repository.

File: bin/env.sh (hunk 3 above)
Comment (cloudxxx8) [reply bin/env.sh]: The same reason as mentioned.

File: bin/env.sh (hunk 2 above)
Comment (cloudxxx8) [reply bin/env.sh]: Thanks for the reminder.  @jinlinGuan will update it to use the same version tag as the ARM one.

File: bin/arm64_env.sh (hunk 1 above)
Comment (lranjbar) [reply bin/arm64_env.sh]: Then it needs to be updated to use the master tag. There is a discrepancy that shouldn't be there.

File: sync.sh [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+     export COMPOSE_FILE="releases/${USE_RELEASE}/compose-files/docker-compose-nexus${USE_REDIS}${USE_NO_SECURITY}${USE_ARM64}.yml"
+else
+     export COMPOSE_FILE="releases/${USE_RELEASE}/compose-files/docker-compose${USE_REDIS}-${USE_RELEASE}${USE_NO_SECURITY}${USE_ARM64}.yml"
+fi
+
+echo $COMPOSE_FILE
+git remote add dev https://github.com/edgexfoundry/developer-scripts.git
+git checkout -b devBranch
+git fetch dev
+git checkout dev/master -- $COMPOSE_FILE
```
Comment (bnevis-i) [inline sync.sh]: Wouldn't it be more straightforward to just 
```
wget "https://raw.githubusercontent.com/edgexfoundry/developer-scripts/master/releases/nightly-build/compose-files/docker-compose-nexus.yml"
```

with a few variable substitutions thrown in?

**Comment by bnevis-i** [comment]
This PR doesn't seem to address the requirement to use kong-centos and kong-ubuntu vs kong-alpine in the CI environment.

File: bin/arm64_env.sh (hunk 1 above)
Comment (jinlinGuan) [reply bin/arm64_env.sh]: fixed

File: bin/env.sh (hunk 2 above)
Comment (jinlinGuan) [reply bin/env.sh]: fixed

File: sync.sh (hunk 4 above)
Comment (bnevis-i) [inline sync.sh]: Suggested replacement:

```
echo $COMPOSE_FILE
wget -q -O docker-compose.yml "https://raw.githubusercontent.com/edgexfoundry/developer-scripts/master/${COMPOSE_FILE}"

if [ `uname -m` = "aarch64" ]; then
    sed -i 's/kong:1.3.0$/kong:1.3.0-ubuntu/' docker-compose.yml
else
    sed -i 's/kong:1.3.0$/kong:1.3.0-centos/' docker-compose.yml
fi
```

**Review by bnevis-i — CHANGES_REQUESTED** [review]

**Comment by bnevis-i** [comment]
`sync.sh` needs to be `chmod +x`

File: deploy-edgeX.sh [hunk 5]
```diff
@@ -25,9 +25,11 @@
 #    . $(dirname "$0")/bin/env.sh
 #fi
 
+source ./sync.sh
```
Comment (bnevis-i) [inline deploy-edgeX.sh]: Suggest just execute
```
./sync.sh
```

and make sync.sh executable.

File: deploy-edgeX.sh [hunk 6]
```diff
@@ -25,9 +25,11 @@
...
+
 run_service () {
 	echo "\033[0;32mStarting.. $1\033[0m"
-	docker-compose up -d $1
+        docker-compose -f $COMPOSE_FILE up -d $1
```
Comment (bnevis-i) [inline deploy-edgeX.sh:33]: Unnecessary change if we just dynamically download `docker-compose.yml` and modify it in place.

Note: be sure to add `docker-compose.yml` to .gitignore.

File: sync.sh (hunk 4 above)
Comment (jinlinGuan) [reply sync.sh]: Why isn't the kong image changed in developer-script instead? This PR [edgexfoundry/developer-scripts#187](https://github.com/edgexfoundry/developer-scripts/pull/187/files) will add docker-compose files which contain kong:1.3.0-ubuntu for arm64.

Comment (bnevis-i) [reply sync.sh]: > Why isn't the kong image changed in developer-script instead? This PR [edgexfoundry/developer-scripts#187](https://github.com/edgexfoundry/developer-scripts/pull/187/files) will add docker-compose files which contain kong:1.3.0-ubuntu for arm64.

Good point.

How about this then:
```
echo $COMPOSE_FILE
wget -q -O docker-compose.yml "https://raw.githubusercontent.com/edgexfoundry/developer-scripts/master/${COMPOSE_FILE}"

# Use Centos base image instead of Alpine base image for Kong for x86_64 CI
# due to compatibility issues with Alpine image in CI
if [ `uname -m` = "x86_64" ]; then
    sed -i 's/kong:1.3.0$/kong:1.3.0-centos/' docker-compose.yml
fi
```

File: deploy-edgeX.sh (hunk 6 above)
Comment (jinlinGuan) [reply deploy-edgeX.sh:33]: The original `docker-compose.yml` is still needed, because it has the services for blackbox-testing only.  Thus, it is necessary to use `$COMPOSE_FILE` to specify the compose file from developer-script.

Comment (bnevis-i) [reply deploy-edgeX.sh:33]: I see now.  Perhaps call them `docker-compose-services-stack.yml` and `docker-compose-test-tools.yml` to distinguish the two?  (This would also eliminate the annoying merge conflicts in the meantime.)

File: sync.sh (hunk 4 above)
Comment (jinlinGuan) [reply sync.sh]: Changed to:
```
wget -q "https://raw.githubusercontent.com/edgexfoundry/developer-scripts/master/releases/${USE_RELEASE}/compose-files/${COMPOSE_FILE}"

# Use Centos base image instead of Alpine base image for Kong for x86_64 CI
# due to compatibility issues with Alpine image in CI
# sed command of MacOS is in different syntax 
[ "$(uname -m)" = "x86_64" ] && [ "$(uname)" = "Darwin" ] && sed -i '' 's/kong:1.3.0$/kong:1.3.0-centos/' ${COMPOSE_FILE} || sed -i 's/kong:1.3.0$/kong:1.3.0-centos/' ${COMPOSE_FILE}

```

File: deploy-edgeX.sh (hunk 5 above)
Comment (jinlinGuan) [reply deploy-edgeX.sh]: fixed

File: deploy-edgeX.sh (hunk 6 above)
Comment (jinlinGuan) [reply deploy-edgeX.sh:33]: Good idea!

The original `docker-compose.yml` is changed to `docker-compose-test-tools.yml`, however, the compose file from developer-scripts still uses the name as its name in developer-scripts to make sure using the correct compose file as expected.

File: bin/appServiceConfigurableTest.sh [hunk 7]
```diff
@@ -19,7 +19,7 @@ echo "Info: Initiating AppServiceConfigurable Test."
 
 echo "[info] ---------- use docker-compose run newman ----------"
 
-docker-compose run --rm postman run ${COLLECTION_PATH} \
+docker-compose -f ../docker-compose-test-tools.yml docker-compose-test-tools.yml run --rm postman run ${COLLECTION_PATH} \
```
Comment (bnevis-i) [inline bin/appServiceConfigurableTest.sh]: Not sure what is going on here with the duplication of `docker-compose-test-tools.yml`.  What is this doing?

File: sync.sh [hunk 8]
```diff
@@ -0,0 +1,31 @@
...
+else
+     COMPOSE_FILE="docker-compose${USE_REDIS}-${USE_RELEASE}${USE_NO_SECURITY}${USE_ARM64}.yml"
+fi
+
+wget -q "https://raw.githubusercontent.com/edgexfoundry/developer-scripts/master/releases/${USE_RELEASE}/compose-files/${COMPOSE_FILE}"
+
+# Use Centos base image instead of Alpine base image for Kong for x86_64 CI
+# due to compatibility issues with Alpine image in CI
+# sed command of MacOS is in different syntax 
+[ "$(uname -m)" = "x86_64" ] && [ "$(uname)" = "Darwin" ] && sed -i '' 's/kong:1.3.0$/kong:1.3.0-centos/' ${COMPOSE_FILE} || sed -i 's/kong:1.3.0$/kong:1.3.0-centos/' ${COMPOSE_FILE}
```
Comment (bnevis-i) [inline sync.sh]: I don't think this line does what you want.  if any of the condtions fails (x86_64, Darwin, or sed) it will end up running the Linux sed command instead of the BSD one.   At a minimum, I would make the x86_64 check into an if/then.

File: bin/appServiceConfigurableTest.sh (hunk 7 above)
Comment (jinlinGuan) [reply bin/appServiceConfigurableTest.sh]: It's a typo. Thank you for catching it.

Fixed.

File: sync.sh (hunk 8 above)
Comment (jinlinGuan) [reply sync.sh]: Changed to:
```
# Use Centos base image instead of Alpine base image for Kong for x86_64 CI
# due to compatibility issues with Alpine image in CI
# sed command of MacOS is in different syntax 
if [ "$(uname -m)" = "x86_64" ] && [ "$(uname)" = "Darwin" ]; then
     sed -i '' 's/kong:1.3.0$/kong:1.3.0-centos/' ${COMPOSE_FILE}
elif [ "$(uname -m)" = "x86_64" ]; then
     sed -i 's/kong:1.3.0$/kong:1.3.0-centos/' ${COMPOSE_FILE}
fi
```

**Comment by cloudxxx8** [comment]
@lranjbar please help review again when you are back.  thanks

**Comment by bnevis-i** [comment]
Blocking issue: docker-edgex-consul was not rebuilt after https://github.com/edgexfoundry/docker-edgex-consul/pull/16 was merged:
 https://nexus3.edgexfoundry.org/#browse/search/docker=format%3Ddocker%20AND%20attributes.docker.imageName%3Ddocker-edgex-consul:912d0fe7b8192392cfe750c17f11e329:a1f3edd8e1b06cfadf6408e1d7903880 (last updated: Thu Dec 12 2019 09:20:57 GMT-0800 (Pacific Standard Time)

File: sync.sh [hunk 9]
```diff
@@ -0,0 +1,33 @@
...
+
+# nightly or other release
+USE_RELEASE=${RELEASE:-nightly-build}
+if [ "$USE_RELEASE" = "nightly-build" ]; then
+     COMPOSE_FILE="docker-compose-nexus${USE_REDIS}${USE_NO_SECURITY}${USE_ARM64}.yml"
+else
+     COMPOSE_FILE="docker-compose${USE_REDIS}-${USE_RELEASE}${USE_NO_SECURITY}${USE_ARM64}.yml"
+fi
+
+wget -q "https://raw.githubusercontent.com/edgexfoundry/developer-scripts/master/releases/${USE_RELEASE}/compose-files/${COMPOSE_FILE}"
```
Comment (bnevis-i) [inline sync.sh]: On my system, without the `-O "${COMPOSE_FILE}"` option, when I re-run `deploy-edgeX.sh`  I get multiple dot versions of the compose file download to my system, which creates problems later.

Comment (jinlinGuan) [reply sync.sh]: added

**Comment by bnevis-i** [comment]
Holding off on further validation until docker-edgex-consul Jenkins pipeline is fixed.  Looks nearly ready to merge.

**Review by bnevis-i — APPROVED** [review]
docker-edgex-consul container was published on Dec 27 and resolves all remaining issues.

IMO this is ready to merge.

LGTM

**Comment by bnevis-i** [comment]
@lranjbar Please re-review.

**Review by lranjbar — APPROVED** [review]
LGTM :+1: Sorry for the delay because of the holidays.

**Comment by cloudxxx8** [comment]
thanks for all you review!
